File: TelerikUIExamples/TelerikUIExamples/CustomAnimationLineChart.m
//
//  CustomAnimationLineChart.m
//  TelerikUIExamples
//
//  Copyright (c) 2013 Telerik. All rights reserved.
//

#import "CustomAnimationLineChart.h"
#import <TelerikUI/TelerikUI.h>

@interface CustomAnimationLineChart() <TKChartDelegate>
@end

@implementation CustomAnimationLineChart
{
    TKChart *_chart;
    BOOL _grow;
}

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        [self addOption:@"Sequential animation" selector:@selector(applySequential)];
        [self addOption:@"Grow animation" selector:@selector(applyGrow)];
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];

    _chart = [[TKChart alloc] initWithFrame:self.view.bounds];
    _chart.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    _chart.allowAnimations = YES;
    _chart.delegate = self;
    [self.view addSubview:_chart];
 
    NSMutableArray *points = [[NSMutableArray alloc] init];
    for (int i = 0; i<10; i++) {
        [points addObject:[[TKChartDataPoint alloc] initWithX:@(i) Y:@(arc4random() % 100)]];
    }
    
    TKChartLineSeries *lineSeries = [[TKChartLineSeries alloc] initWithItems:points];
    CGFloat shapeSize = ([UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPhone) ? 14 : 17;
    lineSeries.style.pointShape = [[TKPredefinedShape alloc] initWithType:TKShapeTypeCircle andSize:CGSizeMake(shapeSize, shapeSize)];
    lineSeries.style.shapeMode = TKChartSeriesStyleShapeModeAlwaysShow;
    lineSeries.selection = TKChartSeriesSelectionDataPoint;
    [_chart addSeries:lineSeries];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)applySequential
{
    _grow = NO;
    [_chart animate];
}

- (void)applyGrow
{
    _grow = YES;
    [_chart animate];
}

#pragma mark TKChartDelegate

 // >> chart-anim-line
- (CAAnimation *)chart:(TKChart *)chart animationForSeries:(TKChartSeries *)series withState:(TKChartSeriesRenderState *)state inRect:(CGRect)rect
{
    CFTimeInterval duration = 0;
    NSMutableArray *animations = [[NSMutableArray alloc] init];
    for (int i = 0; i<state.points.count; i++) {
        
        if (_grow) {
            NSString *keyPath = [NSString stringWithFormat:@"seriesRenderStates.%lu.points.%d.x", (unsigned long)series.index, i];
            TKChartVisualPoint *point = [state.points objectAtIndex:i];
            
            CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:keyPath];
            animation.duration = 0.1 *(i + 0.2);
            animation.fromValue = @0;
            animation.toValue = @(point.x);
            animation.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseOut];
            [animations addObject:animation];
            
            duration = animation.duration;
        }
        else {
            NSString *keyPath = [NSString stringWithFormat:@"seriesRenderStates.%lu.points.%d.y", (unsigned long)series.index, i];
            TKChartVisualPoint *point = [state.points objectAtIndex:i];
            CGFloat oldY = rect.size.height;
            
            if (i > 0) {
                CAKeyframeAnimation *animation = [CAKeyframeAnimation animationWithKeyPath:keyPath];
                animation.duration = 0.1* (i + 1);
                animation.values = @[ @(oldY), @(oldY), @(point.y) ];
                animation.keyTimes = @[ @0, @(i/(i+1.)), @1 ];
                animation.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseOut];
                [animations addObject:animation];
                
                duration = animation.duration;
            }
            else {
                CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:keyPath];
                animation.fromValue = @(oldY);
                animation.toValue = @(point.y);
                animation.duration = 0.1f;
                [animations addObject:animation];
            }
        }
    }
    
    CAAnimationGroup *group = [[CAAnimationGroup alloc] init];
    group.duration = duration;
    group.animations = animations;
    
    return group;
}
 // << chart-anim-line

@end

File: TelerikUIExamplesInSwift/TelerikUIExamplesInSwift/CustomAnimationLineChart.swift
//
//  CustomAnimationLineChart.swift
//  Swift Examples
//
//  Copyright (c) 2014 Telerik. All rights reserved.
//

import Foundation
import QuartzCore

class CustomAnimationLineChart: TKExamplesExampleViewController, TKChartDelegate {
    
    let chart = TKChart()
    var grow = false
    
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: NSBundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)

        self.addOption("Sequential animation", action: applySequential)
        self.addOption("Grow animation", action: applyGrow)
        
        self.selectedOption = 0
    }

    required init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        chart.frame = self.view.bounds
        chart.autoresizingMask = UIViewAutoresizing(rawValue: UIViewAutoresizing.FlexibleWidth.rawValue | UIViewAutoresizing.FlexibleHeight.rawValue)
        chart.allowAnimations = true
        chart.delegate = self
        self.view.addSubview(chart)
        
        var points = [TKChartDataPoint]()
        for i in 0..<10 {
            points.append(TKChartDataPoint(x: i, y: Int(arc4random() % 100)))
        }
        
        let lineSeries = TKChartLineSeries(items: points)
        let shapeSize = CGFloat(UIDevice.currentDevice().userInterfaceIdiom == .Phone ? 14 : 17)
        lineSeries.style.pointShape = TKPredefinedShape(type: TKShapeType.Circle, andSize: CGSizeMake(shapeSize, shapeSize))
        lineSeries.style.shapeMode = TKChartSeriesStyleShapeMode.AlwaysShow
        lineSeries.selection = TKChartSeriesSelection.DataPoint
        chart.addSeries(lineSeries)
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    func applySequential() {
        self.grow = false
        chart.delegate = self
        chart.animate()
    }
    
    func applyGrow() {
        self.grow = true
        chart.delegate = self
        chart.animate()
    }
    
    //MARK: - TKChartDelegate
    
    // >> chart-anim-line-swift
    func chart(chart: TKChart, animationForSeries series: TKChartSeries, withState state: TKChartSeriesRenderState, inRect rect: CGRect) -> CAAnimation? {
        var duration = 0.0
        var animations = [CAAnimation]()
        for i in 0..<state.points.count() {
            if grow {
                let keyPath = "seriesRenderStates.\(series.index).points.\(i).x"
                let point = state.points.objectAtIndex(i) as! TKChartVisualPoint
                let animation = CABasicAnimation(keyPath: keyPath as String)
                animation.duration = 0.1*(Double(i)+0.2)
                animation.fromValue = 0
                animation.toValue = point.x
                animation.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseOut)
                animations.append(animation)
                
                duration = animation.duration
            }
            else {
                let keyPath = "seriesRenderStates.\(series.index).points.\(i).y"
                let point = state.points.objectAtIndex(i) as! TKChartVisualPoint
                let oldY = rect.size.height as CGFloat
                
                if i > 0 {
                    let animation = CAKeyframeAnimation(keyPath: keyPath)
                    animation.duration = 0.1*(Double(i)+1.0)
                    animation.values = [oldY, oldY, point.y]
                    animation.keyTimes = [0, (i/(i+1)), 1]
                    animation.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseOut)
                    animations.append(animation)
                    
                    duration = animation.duration
                }
                else {
                    let animation = CABasicAnimation(keyPath: keyPath)
                    animation.fromValue = oldY
                    animation.toValue = point.y
                    animation.duration = 0.1
                    animations.append(animation)
                }
            }
        }
        
        let group = CAAnimationGroup()
        group.duration = duration
        group.animations = animations
        
        return group
    }
    // << chart-anim-line-swift
}